### PR TITLE
Fix AttributeError in Autocontext when adding data with invalid feature scales (#3106)

### LINKED ISSUE #3106: Autocontext: Can't add additional multiscales with features selected
## Describe the bug
In Autocontext, once the workflow has progressed to feature selection, the "your new image mismatches some selected feature scales" flow is broken because autocontext maintains multiple feature selection applets.

This is a problem because it happens when the "new image" is a multiscale: The default-loaded scale is the smallest, and it's not unusual for that to be too small for already chosen features.

## Expected behavior
The `FeatureSelectionConstraintError.fixing_dialogs[0]` should pop up, but only after the error handling flow that tries to find a matching scale (in case the dataset is multiscale)

## To Reproduce
Autocontext:
1. Add some data in the data selection applet
2. Go to feature selection, select all the features
3. Go back to data selection
4. Export a tiny crop of the original data, like nothing bigger than 10 along spatial axes
5. Try to add the crop in a new lane

Alternatively:
1. Convert some dataset to "multi-scale OME-Zarr"
2. Start Autocontext
3. Add multiscale
4. Switch to raw scale
5. Go to feature selection and select all
6. Go back to data selection
7. Try to add the same multiscale in a new lane

## Error message/traceback
```pytb
ERROR 2025-11-12 09:25:53,233 dataSelectionGui 22544 12388 'AutocontextTwoStage' object has no attribute 'featureSelectionApplet'
Traceback (most recent call last):
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\ilastik\applets\dataSelection\dataSelectionGui.py", line 463, in addLanes
    if not self.applyDatasetInfos(new_infos, info_slots):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\ilastik\applets\dataSelection\dataSelectionGui.py", line 515, in applyDatasetInfos
    raise e
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\ilastik\applets\dataSelection\dataSelectionGui.py", line 499, in applyDatasetInfos
    info_slot.setValue(new_info)
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 73, in call_in_setup_context
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1191, in setValue
    self._changed()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1396, in _changed
    c._changed()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1396, in _changed
    c._changed()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1396, in _changed
    c._changed()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1403, in _changed
    self._configureOperator(self)
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1418, in _configureOperator
    self.operator._setupOutputs()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\operator.py", line 508, in _setupOutputs
    self.setupOutputs()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\ilastik\applets\dataSelection\opDataSelection.py", line 913, in setupOutputs
    self.Image.connect(op5.Output)
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 73, in call_in_setup_context
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 562, in connect
    self._changed()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1396, in _changed
    c._changed()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1396, in _changed
    c._changed()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1396, in _changed
    c._changed()
  [Previous line repeated 2 more times]
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1403, in _changed
    self._configureOperator(self)
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\slot.py", line 1418, in _configureOperator
    self.operator._setupOutputs()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\lazyflow\operator.py", line 508, in _setupOutputs
    self.setupOutputs()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\ilastik\applets\featureSelection\opFeatureSelection.py", line 253, in setupOutputs
    super().setupOutputs()
  File "C:\Program Files\ilastik-1.4.2b5\Lib\site-packages\ilastik\applets\featureSelection\opFeatureSelection.py", line 172, in setupOutputs
    if self.parent.parent.featureSelectionApplet._gui is None:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'AutocontextTwoStage' object has no attribute 'featureSelectionApplet'
```

## Desktop (please complete the following information):
 - ilastik version: confirmed in 1.4.1post1 and 1.4.2b5
 - OS: win11 but should be independent

### PR DESCRIPTION
**Pull request by rajdhruvsingh**
## Summary
Fixes a crash in the Autocontext (2-stage) workflow when adding a dataset
whose dimensions are too small for the selected feature scales.

Fixes #3106

## Root Cause
`OpFeatureSelectionNoCache.setupOutputs()` accesses `self.parent.parent.featureSelectionApplet`
to show the invalid scales error dialog. Regular workflows expose a single
`featureSelectionApplet` attribute, but Autocontext workflows expose
`featureSelectionApplets` (plural, a list). The code assumed the singular
form always existed, causing:
AttributeError: 'AutocontextTwoStage' object has no attribute 'featureSelectionApplet'
## Fix
Use `getattr` with fallback to find the correct feature selection applet —
first try `featureSelectionApplet` (singular), then fall back to
`featureSelectionApplets` (plural list), finding the correct applet by
matching `topLevelOperator`:

```python
workflow = self.parent.parent
feature_applet = getattr(workflow, "featureSelectionApplet", None)
if feature_applet is None:
    feature_applets = getattr(workflow, "featureSelectionApplets", [])
    feature_applet = next(
        (a for a in feature_applets if a.topLevelOperator is self.parent),
        feature_applets[0] if feature_applets else None,
    )
```

## Before / After

**Before:** Raw `AttributeError` dialog with no actionable information.

**After:** Proper `FeatureSelectionConstraintError` dialog telling the user
which scales are too large and what to do.

## Changes
- `ilastik/applets/featureSelection/opFeatureSelection.py` — fix applet lookup
- `tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py` — 3 tests documenting the lookup logic

## Checklist
- [x] Format code and imports.
- [x] Add docstrings and comments.
- [x] Add tests.
- [x] Reference relevant issues and other pull requests.
- [ ] Rebase commits into a logical sequence.
- [ ] Update user documentation. (Not required - bug fix restoring expected behavior)
- [x] Add screenshots / screen recordings.

## Testing
All 6 tests pass:
TestOpFeatureSelection::testDirtyPropagation PASSED

TestOpFeatureSelection::test_2d PASSED

TestOpFeatureSelection::test_basicFunctionality PASSED

TestOpFeatureSelectionAppletLookup::test_finds_applet_in_autocontext_workflow PASSED

TestOpFeatureSelectionAppletLookup::test_finds_applet_in_regular_workflow PASSED

TestOpFeatureSelectionAppletLookup::test_returns_none_when_no_applet_found PASSED

Screenshots of before/after behavior attached below:
Screenshot 2026-06-14 at 23 25 24

preview

### CODE REVIEW

**Review by k-dominik — CHANGES_REQUESTED** [review]
Hi @rajdhruvsingh,

first of all, thank you for putting in the time to work on this! Great that you have included screenshots!

We have a pull request template that includes a checklist. Please don't delete the checklist (and edit the PR description to include the checklist). We are less likely to look into PRs with missing checklist in detail.

**Comment by rajdhruvsingh** [comment]
Thanks @k-dominik! I've updated the PR description and added the checklist.

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 1]
```diff
@@ -163,3 +163,71 @@ def handleDirty(slot, roi):
...
+        mock_applet = MagicMock()
+        mock_workflow = MagicMock(spec=["featureSelectionApplet"])
+        mock_workflow.featureSelectionApplet = mock_applet
+
+        feature_applet = getattr(mock_workflow, "featureSelectionApplet", None)
+        if feature_applet is None:
+            feature_applets = getattr(mock_workflow, "featureSelectionApplets", [])
+            feature_applet = next(
+                (a for a in feature_applets if a.topLevelOperator is None),
+                feature_applets[0] if feature_applets else None,
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: You're testing the code by mirroring the implementation in the test. This style of testing has some important drawbacks: The most severe one being that it doesn't guarantee that correctness of the code in `opFeatureSelection` - because it does not test anything from there.

Line 187 is always false - in line 184 you set the `featureSelectionApplet` attribute as `mock_applet` so it's never None.

**Review by k-dominik — CHANGES_REQUESTED** [review]
The change is  fine, the tests, however, are not. None of the tests is testing the actual code from `opFeatureSelection.py` but rather mirror the implementation in some ways. This is too hard to maintain / keep in sync.

**Comment by rajdhruvsingh** [comment]
Hi @k-dominik, You're right — the previous tests were mirroring the implementation rather than calling into it. I've replaced them with tests that instantiate a real OpFeatureSelectionNoCache, patch parent.parent to point to a mock workflow, and trigger the invalid-scales path by setting a sigma=10.0 feature on a 5×5 image — forcing setupOutputs() in opFeatureSelection.py to actually execute the applet-lookup code. Each test asserts on ctx.exception.fixing_dialogs to verify the correct applet was found.

I also verified this manually in the Autocontext (2-stage) workflow — adding a small dataset with large scales selected now shows the proper FeatureSelectionConstraintError dialog (screenshot attached) instead of crashing with AttributeError.

To confirm the tests catch real regressions: temporarily reverting the getattr fix to a direct attribute access caused 2 of the 3 new tests to fail as expected.

Screenshot 2026-08-01 at 18 59 23

**Comment by rajdhruvsingh** [comment]
Hi @k-dominik, a follow up on this PR would be really helpful. I hope you'll find time to look through it.
Thank You

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 2]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
...
+
+        We patch the operator's parent property after all slots are configured,
+        because lazyflow calls setupOutputs() immediately on setValue() and the
+        parent chain must be in place by then.
+        """
+        from unittest.mock import MagicMock, patch
+        from ilastik.applets.featureSelection.opFeatureSelection import OpFeatureSelectionNoCache
+        import numpy
+        import vigra
+        from lazyflow.graph import Graph
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: Please move the imports to the top of the file.

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 3]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
...
+
+        graph = Graph()
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: Minor, but there is also a `graph` fixture you can use (we try to use it in new tests)

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 4]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
...
+        op._test_inner_mock = inner_mock
+        OpFeatureSelectionNoCache.parent = property(lambda self: inner_mock)
+        op._test_selections = numpy.zeros((6, 7), dtype=bool)
+        op._test_selections[:, 6] = True  # sigma=10.0, invalid for 5x5 image
+        return op
+
+    def tearDown(self):
+        # Restore the original parent property after each test
+        from ilastik.applets.featureSelection.opFeatureSelection import OpFeatureSelectionNoCache
+        from lazyflow.operator import Operator
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: Please move to top of the file

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 5]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
...
+
+        OpFeatureSelectionNoCache.parent = Operator.parent
+
+    def test_regular_workflow_uses_singular_applet(self):
+        """
+        In a regular workflow, setupOutputs() finds featureSelectionApplet
+        (singular) and includes its callback in fixing_dialogs.
+        """
+        from unittest.mock import MagicMock
+        from ilastik.applets.featureSelection.opFeatureSelection import FeatureSelectionConstraintError
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: Please move to top of the file

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 6]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
...
+        self.assertIn(mock_callback, ctx.exception.fixing_dialogs)
+
+    def test_autocontext_workflow_uses_plural_applets(self):
+        """
+        In an Autocontext workflow, setupOutputs() falls back to
+        featureSelectionApplets (plural) and picks the applet whose
+        topLevelOperator matches self.parent.
+        """
+        from unittest.mock import MagicMock
+        from ilastik.applets.featureSelection.opFeatureSelection import FeatureSelectionConstraintError
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: Please move to top of the file

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 7]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
...
+        self.assertIn(mock_callback_0, ctx.exception.fixing_dialogs)
+
+    def test_headless_produces_empty_fix_dlgs(self):
+        """
+        When no applet is found (headless or unknown workflow), setupOutputs()
+        raises FeatureSelectionConstraintError with an empty fixing_dialogs
+        rather than crashing with AttributeError.
+        """
+        from unittest.mock import MagicMock
+        from ilastik.applets.featureSelection.opFeatureSelection import FeatureSelectionConstraintError
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: Please move to top of the file

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 8]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
...
+            op.FeatureIds.setValue(
+                [
+                    "GaussianSmoothing",
+                    "LaplacianOfGaussian",
+                    "StructureTensorEigenvalues",
+                    "HessianOfGaussianEigenvalues",
+                    "GaussianGradientMagnitude",
+                    "DifferenceOfGaussians",
+                ]
+            )
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: `Scales`, and `FeatureIds` have default values that are identical to the ones you're setting here. Is there a reason for this?

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 9]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
...
+        # Re-attach the patch permanently so setupOutputs() can access it
+        # when called explicitly in each test.
+        op._test_inner_mock = inner_mock
+        OpFeatureSelectionNoCache.parent = property(lambda self: inner_mock)
+        op._test_selections = numpy.zeros((6, 7), dtype=bool)
+        op._test_selections[:, 6] = True  # sigma=10.0, invalid for 5x5 image
+        return op
+
+    def tearDown(self):
+        # Restore the original parent property after each test
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: teardown should not be necessary - can you confirm you run into issues without it? `patch.object` contextmanager should take care of that.

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 10]
```diff
@@ -163,3 +163,140 @@ def handleDirty(slot, roi):
         assert (dirtyRois[0].start, dirtyRois[0].stop) == sliceToRoi(
             slice(None), self.opFeatures.OutputImage[0].meta.shape
         )
+
+
+class TestOpFeatureSelectionAppletLookup(unittest.TestCase):
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: please don't create tests as unittest sublcasses. We've migrated to pytest (which in principle also supports setup and teardown if needed).

**Comment by rajdhruvsingh** [comment]
Hi @k-dominik , is there anything still left from me to fix that might still be blocking the merge of this PR?

**Review by k-dominik — CHANGES_REQUESTED** [review]
The new tests work and test the correct behavior. They (re) introduce tests based on `unittest.TestCase` though, which we do not do for new tests (and should probably document this). So you can in fact make use of the whole pytest machinery.

Just for inspiration, I would rewrite the tests something like:

```Python
class TestOpFeatureSelectionAppletLookup:
    """
    Tests that OpFeatureSelectionNoCache.setupOutputs() correctly finds the
    feature selection applet in both regular and Autocontext workflows by
    actually calling into the production code in opFeatureSelection.py.

    Fixes #3106: 'AutocontextTwoStage' object has no attribute 'featureSelectionApplet'
    """

    @pytest.fixture
    def invalid_selection(self):
        test_selections = numpy.zeros((6, 7), dtype=bool)
        test_selections[:, 6] = True  # sigma=10.0, invalid for 5x5 image
        return test_selections

    @pytest.fixture
    def mock_workflow(self):
        return MagicMock()

        @pytest.fixture
    def op_with_invalid_scales(self, mock_workflow):
        """
        Create a real OpFeatureSelectionNoCache whose parent.parent resolves
        to mock_workflow, configured so setupOutputs() hits the invalid-scales
        branch and exercises the applet-lookup code in production.

        We patch the operator's parent property after all slots are configured,
        because lazyflow calls setupOutputs() immediately on setValue() and the
        parent chain must be in place by then.
        """
        graph = Graph()

        # Build the inner mock (op.parent) whose .parent is mock_workflow
        inner_mock = MagicMock()
        inner_mock.parent = mock_workflow

        with patch.object(
            OpFeatureSelectionNoCache,
            "parent",
            new_callable=PropertyMock,
        ) as mock:
            mock.return_value = inner_mock
            op = OpFeatureSelectionNoCache(graph=graph)

            # Tiny image: too small for large sigma scales -> triggers invalid_scales path
            tiny = numpy.zeros((1, 5, 5, 1, 1), dtype=numpy.float32)
            tiny = vigra.taggedView(tiny, "txyzc")
            op.InputImage.setValue(tiny)
            yield op

    def test_regular_workflow_uses_singular_applet(
        self,
        mock_workflow: MagicMock,
        op_with_invalid_scales: OpFeatureSelectionNoCache,
        invalid_selection: numpy.ndarray,
    ):
        """
        In a regular workflow, setupOutputs() finds featureSelectionApplet
        (singular) and includes its callback in fixing_dialogs.
        """
        mock_callback = MagicMock()
        mock_applet = MagicMock()
        mock_applet._gui.currentGui.return_value.onFeatureButtonClicked = mock_callback

        mock_workflow.featureSelectionApplet = mock_applet

        with pytest.raises(FeatureSelectionConstraintError) as ex:
            op_with_invalid_scales.SelectionMatrix.setValue(invalid_selection)

            assert mock_callback in ex.exception.fixing_dialogs

    def test_autocontext_workflow_uses_plural_applets(
        self,
        mock_workflow: MagicMock,
        op_with_invalid_scales: OpFeatureSelectionNoCache,
        invalid_selection: numpy.ndarray,
    ):
        """
        In an Autocontext workflow, setupOutputs() falls back to
        featureSelectionApplets (plural) and picks the applet whose
        topLevelOperator matches self.parent.
        """
        mock_callback_0 = MagicMock()
        mock_applet_0 = MagicMock()
        mock_applet_0._gui.currentGui.return_value.onFeatureButtonClicked = mock_callback_0
        mock_applet_1 = MagicMock()

        mock_workflow.featureSelectionApplets = [mock_applet_0, mock_applet_1]

        # Match applet_0 to op.parent so production code picks it
        mock_applet_0.topLevelOperator = op_with_invalid_scales.parent

        with pytest.raises(FeatureSelectionConstraintError) as ex:
            op_with_invalid_scales.SelectionMatrix.setValue(invalid_selection)
            assert mock_callback_0 in ex.fixing_dialogs

... for headless test something similar
```

**Comment by rajdhruvsingh** [comment]
Hi @k-dominik, Thanks for the detailed feedback and the example — very helpful. I've addressed all the comments:

Converted to a plain pytest class (no unittest.TestCase)
Moved all imports to the top of the file
Switched to PropertyMock via patch.object context manager (no manual tearDown needed)
Removed the local graph fixture and used the global one from tests/conftest.py instead
Removed the redundant Scales and FeatureIds setValue calls since the defaults are identical
Fixed the assert placement — moved outside the with pytest.raises block so it actually executes

I also re-verified that the tests catch real regressions: temporarily replacing getattr(workflow, "featureSelectionApplet", None) with a direct attribute access caused the autocontext and headless tests to fail as expected.

**Comment by rajdhruvsingh** [comment]
Hi @k-dominik, a follow up on this PR would be really helpful. I hope you'll find time to look through it.
Thank You

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 11]
```diff
@@ -163,3 +170,115 @@ def handleDirty(slot, roi):
...
+        op_with_invalid_scales,
+        invalid_selection,
+    ):
+        """
+        When no applet is found (headless or unknown workflow), setupOutputs()
+        raises FeatureSelectionConstraintError with an empty fixing_dialogs
+        rather than crashing with AttributeError.
+        """
+        del mock_workflow.featureSelectionApplet
+        del mock_workflow.featureSelectionApplets
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py]: in headless the code `feature_applet._gui` is unset - this is not what you're testing in this function. While the result is the same, it doesn't test the underlying assumption. Please adapt the test such that it relies on an unset _gui atttribute instead.

**Review by k-dominik — APPROVED** [review]
Hi @rajdhruvsingh,

thank you for incorporating the changes! There's only one thing left (see inline comment). And also in https://github.com/ilastik/ilastik/pull/3227#pullrequestreview-4873876413 I added some type annotations for a reason: to allow language analysis tools to infer the types of fixtures correctly - so please add those as in the comment.

**Review by k-dominik — CHANGES_REQUESTED** [review]
Sorry - status is still in "request changes"

**Comment by rajdhruvsingh** [comment]
Hi @k-dominik, I have Fixed the headless test now sets mock_applet._gui = None to test the feature_applet._gui is None branch in production. In ilastik's headless initialization, _gui is explicitly set to None rather than being absent, so this correctly reflects the real headless state. I've updated the docstring to make this explicit.

Also addressed the other feedback: added type annotations to all fixture return types and test method parameters, added a local graph fixture with return type, and merged the duplicate mock imports into one line.

Regression check confirms the tests catch real breakage: reverting getattr to direct attribute access causes the autocontext test to fail as expected.

File: tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py [hunk 12]
```diff
@@ -163,3 +169,120 @@ def handleDirty(slot, roi):
...
+    feature selection applet in both regular and Autocontext workflows by
+    actually calling into the production code in opFeatureSelection.py.
+
+    Fixes #3106: 'AutocontextTwoStage' object has no attribute 'featureSelectionApplet'
+    """
+
+    @pytest.fixture
+    def graph(self) -> Graph:
+        return Graph()
+
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/featureSelection/testOpFeatureSelection.py:186]:
```suggestion
```

the graph fixture is available in top-level conftest. I admit, not super discoverable, but would want to avoid having it here again.

**Review by k-dominik — CHANGES_REQUESTED** [review]
Looks good except for the superfluous graph fixture :). Please remove only that and we're good to go.

**Comment by rajdhruvsingh** [comment]
Hi @k-dominik, I have removed the local graph fixture. The project-wide one in tests/conftest.py is picked up automatically and all 6 tests pass.

**Review by k-dominik — APPROVED** [review]
Hey @rajdhruvsingh,

thank you so much for bearing with me during this admittedly lengthy review process! Changes look good now and can go in!

Thank you for your contribution!

Cheers
Dominik
